### PR TITLE
Migrate to View binding, SDK 34 and Version catalogs

### PR DESCRIPTION
**Pull request by naveensingh**
### Major changes:
 - Migrate build scripts from Groovy to [Kotlin DSL](https://developer.android.com/build/migrate-to-kotlin-dsl)
  - Migrate to [Gradle version catalogs](https://developer.android.com/build/migrate-to-catalogs)
 - Migrate from Kotlin Kapt to [KSP (Kotlin symbol processing)](https://developer.android.com/build/migrate-to-ksp) as Kapt is now in maintenance mode.
 - Migrate from Kotlin synthetics to [View binding](https://developer.android.com/topic/libraries/view-binding)
 - Target SDK 34:
   - Add `FOREGROUND_SERVICE_MEDIA_PLAYBACK` permission
   - No other changes are required. I tested using an Android 14 emulator.

### Minor changes: 
 - Close track activity if the track is not found on third-party intents. Related issue: The commons function `getRealPathFromUri` used at third-party intents isn't working for files with download documents uri, I'll check it soon.
 - Fixed some minor adapter-related issues.
 - Add proguard rules for Gson to avoid `Fatal Exception: java.lang.IllegalStateException: TypeToken must be created with a type argument`. See https://r8.googlesource.com/r8/+/refs/heads/main/compatibility-faq.md#troubleshooting-gson-gson. **Why do we need Gson here anyway? We can just do `Intent.putParcelable()`**, once I implemented an app using intent parcelables and it worked pretty well.
 - Remove Picasso proguard rules.
 - Fixed a repeat mode glitch caused by last-minute changes in the media3 PR. 

### Notes:
 - Since we need access to the currently initialized fragments in the main activity, I simply used the PagerAdapter API to store initialized fragments in a list. I didn't look for a novel solution since all this will be rewritten but let me know if I should.

### CODE REVIEW

File: app/src/main/kotlin/com/simplemobiletools/musicplayer/activities/MainActivity.kt [hunk 1]
```diff
@@ -1,5 +1,6 @@
 package com.simplemobiletools.musicplayer.activities
 
+
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/musicplayer/activities/MainActivity.kt:3]: useless empty line

File: app/src/main/kotlin/com/simplemobiletools/musicplayer/activities/QueueActivity.kt [hunk 2]
```diff
@@ -8,40 +8,43 @@ import android.view.MenuItem
...
 import com.simplemobiletools.commons.helpers.ensureBackgroundThread
 import com.simplemobiletools.musicplayer.R
 import com.simplemobiletools.musicplayer.adapters.QueueAdapter
+import com.simplemobiletools.musicplayer.databinding.ActivityQueueBinding
 import com.simplemobiletools.musicplayer.dialogs.NewPlaylistDialog
 import com.simplemobiletools.musicplayer.extensions.*
 import com.simplemobiletools.musicplayer.helpers.RoomHelper
 import com.simplemobiletools.musicplayer.models.Track
-import kotlinx.android.synthetic.main.activity_queue.*
+
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/musicplayer/activities/QueueActivity.kt:24]: another useless line

File: app/src/main/kotlin/com/simplemobiletools/musicplayer/fragments/PlaylistsFragment.kt [hunk 3]
```diff
@@ -1,39 +1,38 @@
 package com.simplemobiletools.musicplayer.fragments
 
+
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/musicplayer/fragments/PlaylistsFragment.kt:3]: another line :)

File: app/src/main/kotlin/com/simplemobiletools/musicplayer/playback/MediaSessionCallback.kt [hunk 4]
```diff
@@ -14,6 +14,7 @@ import com.simplemobiletools.musicplayer.extensions.addRemainingMediaItems
 import com.simplemobiletools.musicplayer.extensions.config
 import com.simplemobiletools.musicplayer.helpers.EXTRA_NEXT_MEDIA_ID
 import com.simplemobiletools.musicplayer.helpers.EXTRA_SHUFFLE_INDICES
+import com.simplemobiletools.musicplayer.helpers.log
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/musicplayer/playback/MediaSessionCallback.kt:17]: accidental import

**Comment by tibbi** [comment]
have you also tested upgrading from master branch in release mode?

**Comment by tibbi** [comment]
selecting items like folders by long press and drag doesnt work

**Comment by naveensingh** [comment]
> have you also tested upgrading from master branch in release mode?

No, I did some tests in release mode but not the migration one. Only tested the upgrade with debug builds, will check it now.
I'd tested migration in release mode at the time of media3 changes though

**Comment by naveensingh** [comment]
> selecting items like folders by long press and drag doesnt work, only works at main tracks fragment

I'm on it...

**Comment by naveensingh** [comment]
> selecting items like folders by long press and drag doesnt work, only works at main tracks fragment

I broke it :)

**Comment by naveensingh** [comment]
> have you also tested upgrading from master branch in release mode?

did some basic testing, and nothing looks broken when upgrading from the last release in release mode

**Comment by naveensingh** [comment]
Fix everything, check now, please.

**Comment by tibbi** [comment]
if I have 1 playlist only and I long press it, it doesnt have the selected background. Albums opened from Artists tab dont have selected backgrounds either.

**Comment by naveensingh** [comment]
> I have 1 playlist only and I long press it, it doesnt have the selected background

have you pulled the latest changes? I broke it during media3 changes and fixed it above https://github.com/SimpleMobileTools/Simple-Music-Player/pull/592/commits/bebd99bf5ff53a77ae0789e746e13fd10a92cf52

> Albums opened from Artists tab dont have selected backgrounds either

I remember that one never really worked and I kept ignoring it, I'm going to fix it now.

**Comment by tibbi** [comment]
right, it works well now, but colors arent properly updated in onResume. The placeholder icon and selectiong texts either.

**Comment by naveensingh** [comment]
Not sure what I need to do reproduce this.  I select some items   => go to home screen => reopen the app => everything is just as before

**Comment by tibbi** [comment]
no need to select anything, just open the app settings, change colors and go back at the main screen

**Comment by naveensingh** [comment]
Fixed the color issues and resolved some conflicts. I think the colors were not updated in some other apps with lists, will have to check.

**Comment by naveensingh** [comment]
I didn't have any tracks without placeholders so didn't notice it, added now.

**Comment by tibbi** [comment]
yeah looks good now, lets see how will it go ;) Thanks

**Review by GregKruz — CHANGES_REQUESTED** [review]
Spam

**Pull request merged**